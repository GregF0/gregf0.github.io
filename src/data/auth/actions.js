import { firebase } from '../../firebase/firebase'

export const setUserInfo = (userInfo) => ({
  type: 'SET_USER_INFO',
  userInfo,
})

export const linkAccount = (authProvider) => {
  return async (dispatch) => {
    try {
      const res = await firebase.auth().currentUser.linkWithPopup(authProvider)

      const user = res.user

      const userInfo = { uid: user.uid }

      userInfo.creationTime = user.metadata.creationTime

      if (user.providerData && user.providerData.length) {
        userInfo.name = user.providerData[0].displayName
        userInfo.photo = user.providerData[0].photoURL
      }

      dispatch(setUserInfo(userInfo))
      return res
    } catch (e) {
      if (e.code === 'auth/credential-already-in-use') {
        return firebase.auth().signInWithCredential(e.credential)
      }
    }
  }
}

export const signOut = () => ({
  type: 'SIGN_OUT',
})

export const startSignOut = () => {
  return async (dispatch) => {
    await firebase.auth().signOut()
    dispatch(signOut())
  }
}
