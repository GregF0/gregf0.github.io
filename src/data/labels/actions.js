import fs from '../../firebase/firebase'

export const setFullscreenDialog = (fullscreenDialog) => ({
  type: 'SET_FULLSCREEN_DIALOG',
  fullscreenDialog,
})

export const setDesktopDialog = (desktopDialog) => ({
  type: 'SET_DESKTOP_DIALOG',
  desktopDialog,
})

export const setDeleteAlert = (deleteAlert) => ({
  type: 'SET_DELETE_ALERT',
  deleteAlert,
})

export const setMenuOpened = (menuOpened) => ({
  type: 'SET_MENU_OPENED',
  menuOpened,
})

export const setFormValue = (formValue) => ({
  type: 'SET_FORM_VALUE',
  formValue,
})

export const setLabelEditting = (labelEditting) => ({
  type: 'SET_LABEL_EDITTING',
  labelEditting,
})

export const setLabelSelected = (labelSelected) => ({
  type: 'SET_LABEL_SELECTED',
  labelSelected,
})

export const addLabel = (label) => ({
  type: 'ADD_LABEL',
  label,
})

export const startAddLabel = (label) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    const newLabelRef = fs.collection(`users/${uid}/labels`).doc()

    dispatch(
      addLabel({
        id: newLabelRef.id,
        ...label,
      })
    )
    await newLabelRef.set({ ...label })
  }
}

export const editLabel = (id, updates) => ({
  type: 'EDIT_LABEL',
  id,
  updates,
})

export const startEditLabel = (id, updates) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(editLabel(id, updates))

    await fs.doc(`users/${uid}/labels/${id}`).update({
      ...updates,
    })
  }
}

export const deleteLabel = (id) => ({
  type: 'DELETE_LABEL',
  id,
})

export const startDeleteLabel = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(deleteLabel(id))

    await fs.doc(`users/${uid}/labels/${id}`).delete()
  }
}

export const setLabels = (labels) => ({
  type: 'SET_LABELS',
  labels,
})

export const startSetLabels = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    try {
      const labelsRef = await fs.collection(`users/${uid}/labels`).get()

      const labels = labelsRef.docs.map((label) => ({
        id: label.id,
        ...label.data(),
      }))

      dispatch(setLabels(labels))

      return labelsRef
    } catch (e) {
      dispatch(setLabels([]))
    }
  }
}
