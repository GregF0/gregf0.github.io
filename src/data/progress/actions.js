export const setProgressVisibility = (visibility) => ({
  type: 'SET_PROGRESS_VISIBILITY',
  visibility,
})
