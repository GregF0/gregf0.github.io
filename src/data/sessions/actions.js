import fs from '../../firebase/firebase'

export const setSessions = (sessions) => ({
  type: 'SET_SESSIONS',
  sessions,
})

export const startSetSessions = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    try {
      const sessionsRef = await fs.collection(`users/${uid}/sessions`).get()

      const sessions = sessionsRef.docs.map((session) => ({
        id: session.id,
        ...session.data(),
      }))

      dispatch(setSessions(sessions))
      return sessionsRef
    } catch (e) {
      dispatch(setSessions([]))
    }
  }
}

export const addSession = (session) => ({
  type: 'ADD_SESSION',
  session,
})

export const startAddSession = (session) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid
    const { duration, label = null, createdAt } = session

    const newSessionRef = fs.collection(`users/${uid}/sessions`).doc()

    dispatch(
      addSession({
        id: newSessionRef.id,
        ...session,
      })
    )

    await newSessionRef.set({
      duration,
      label,
      createdAt,
    })

    return newSessionRef
  }
}
