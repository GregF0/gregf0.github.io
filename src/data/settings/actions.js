import fs from '../../firebase/firebase'

export const setWorkDuration = (duration) => ({
  type: 'SET_WORK_DURATION',
  duration,
})

export const startSetWorkDuration = (duration) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setWorkDuration(duration))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { workDuration: duration } }, { merge: true })
  }
}

export const setShortBreakDuration = (duration) => ({
  type: 'SET_SHORT_BREAK_DURATION',
  duration,
})

export const startSetShortBreakDuration = (duration) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setShortBreakDuration(duration))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { shortBreakDuration: duration } }, { merge: true })
  }
}

export const setLongBreakDuration = (duration) => ({
  type: 'SET_LONG_BREAK_DURATION',
  duration,
})

export const startSetLongBreakDuration = (duration) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setLongBreakDuration(duration))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { longBreakDuration: duration } }, { merge: true })
  }
}

export const setRounds = (rounds) => ({
  type: 'SET_ROUNDS',
  rounds,
})

export const startSetRounds = (rounds) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setRounds(rounds))

    await fs.doc(`users/${uid}`).set({ settings: { rounds } }, { merge: true })
  }
}

export const setShowTimerInTitle = (showTimerInTitle) => ({
  type: 'SET_SHOW_TIMER_IN_TITLE',
  showTimerInTitle,
})

export const startSetShowTimerInTitle = (showTimerInTitle) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setShowTimerInTitle(showTimerInTitle))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { showTimerInTitle } }, { merge: true })
  }
}

export const setShowNotifications = (showNotifications) => ({
  type: 'SET_SHOW_NOTIFICATIONS',
  showNotifications,
})

export const startSetShowNotifications = (showNotifications) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setShowNotifications(showNotifications))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { showNotifications } }, { merge: true })
  }
}

export const setDarkMode = (darkMode) => ({
  type: 'SET_DARK_MODE',
  darkMode,
})

export const startSetDarkMode = (darkMode) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    localStorage.setItem('darkMode', +darkMode)

    dispatch(setDarkMode(darkMode))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { darkMode } }, { merge: true })
  }
}

export const setAutostart = (autostart) => ({
  type: 'SET_AUTOSTART',
  autostart,
})

export const startSetAutostart = (autostart) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setAutostart(autostart))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { autostart } }, { merge: true })
  }
}

export const setFirstDayOfTheWeek = (firstDayOfTheWeek) => ({
  type: 'SET_FIRST_DAY_OF_THE_WEEK',
  firstDayOfTheWeek,
})

export const startSetFirstDayOfTheWeek = (firstDayOfTheWeek) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setFirstDayOfTheWeek(firstDayOfTheWeek))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { firstDayOfTheWeek } }, { merge: true })
  }
}

export const setSettings = (settings) => ({
  type: 'SET_SETTINGS',
  settings,
})

export const startSetSettings = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    try {
      const docRef = await fs.doc(`users/${uid}`).get()
      let data = {}

      if (docRef.exists) {
        data = docRef.data().settings
      }

      dispatch(setSettings(data))
    } catch (e) {
      dispatch(setSettings({}))
    }
  }
}
