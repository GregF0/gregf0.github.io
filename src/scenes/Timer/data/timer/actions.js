export const setStatus = (status) => ({
  type: 'SET_STATUS',
  status,
})

export const setProgress = (progress) => ({
  type: 'SET_PROGRESS',
  progress,
})

export const setTimeLeft = (timeLeft) => ({
  type: 'SET_TIME_LEFT',
  timeLeft,
})

export const saveInterval = (interval) => ({
  type: 'SAVE_INTERVAL',
  interval,
})

export const pauseTimer = () => ({
  type: 'PAUSE_TIMER',
})

export const resetTimer = ({ duration, showTimerInTitle }) => ({
  type: 'RESET_TIMER',
  duration,
  showTimerInTitle,
})

export const setNextTimer = (settings) => ({
  type: 'SET_NEXT_TIMER',
  settings,
})

export const setSaveSessionAlert = (saveSessionAlert) => ({
  type: 'SET_SAVE_SESSION_ALERT',
  saveSessionAlert,
})
