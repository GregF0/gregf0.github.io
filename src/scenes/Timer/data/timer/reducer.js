export const STATUSES = {
  onHold: 'ON_HOLD',
  running: 'RUNNING',
  paused: 'PAUSED',
}

export const TYPES = {
  work: 'WORK',
  shortBreak: 'SHORT_BREAK',
  longBreak: 'LONG_BREAK',
}

export const initialState = {
  status: STATUSES.onHold,
  type: TYPES.work,
  progress: 100,
  timeLeft: null,
  currentRound: 1,
  interval: null,
  saveSessionAlert: false,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_STATUS':
      return {
        ...state,
        status: action.status,
      }
    case 'SET_PROGRESS':
      return {
        ...state,
        progress: action.progress,
      }
    case 'SET_TIME_LEFT':
      const { minutes, seconds } = action.timeLeft

      return {
        ...state,
        timeLeft: {
          minutes: minutes,
          seconds: isNaN(seconds) ? 0 : seconds,
        },
      }

    case 'SAVE_INTERVAL':
      return {
        ...state,
        interval: action.interval,
      }
    case 'PAUSE_TIMER':
      clearInterval(state.interval)

      return {
        ...state,
        status: STATUSES.paused,
        interval: null,
      }
    case 'RESET_TIMER':
      clearInterval(state.interval)
      const timeLeft = { minutes: action.duration, seconds: 0 }

      if (action.showTimerInTitle) {
        setTitle(state.type, timeLeft)
      }

      return {
        ...state,
        status: STATUSES.onHold,
        interval: null,
        timeLeft,
        progress: 100,
      }
    case 'SET_NEXT_TIMER':
      clearInterval(state.interval)

      const {
        workDuration,
        shortBreakDuration,
        longBreakDuration,
        rounds,
        showTimerInTitle,
      } = action.settings

      let newType
      let newTimeLeft
      let newCurrentRound = state.currentRound

      if (state.currentRound < rounds) {
        if (state.type === TYPES.work) {
          newType = TYPES.shortBreak
          newTimeLeft = { minutes: shortBreakDuration, seconds: 0 }
        } else {
          newType = TYPES.work
          newTimeLeft = { minutes: workDuration, seconds: 0 }
          newCurrentRound = state.currentRound + 1
        }
      } else {
        if (state.type === TYPES.work) {
          newType = TYPES.longBreak
          newTimeLeft = { minutes: longBreakDuration, seconds: 0 }
        } else {
          newType = TYPES.work
          newTimeLeft = { minutes: workDuration, seconds: 0 }
          newCurrentRound = 1
        }
      }

      if (showTimerInTitle) {
        setTitle(newType, newTimeLeft)
      }

      return {
        ...state,
        status: STATUSES.onHold,
        progress: 100,
        interval: null,
        type: newType,
        timeLeft: newTimeLeft,
        currentRound: newCurrentRound,
      }

    case 'SET_SAVE_SESSION_ALERT':
      return {
        ...state,
        saveSessionAlert: action.saveSessionAlert,
      }
    default:
      return state
  }
}

export const setTitle = (type, timeLeft) => {
  let emoji

  if (type === TYPES.work) {
    emoji = 'üë®‚Äçüíª'
  } else {
    emoji = '‚òïÔ∏è'
  }

  document.title = `${
    timeLeft.minutes < 10 ? '0' + timeLeft.minutes : timeLeft.minutes
  }:${
    timeLeft.seconds < 10 ? '0' + timeLeft.seconds : timeLeft.seconds
  } ${emoji} | Pomodor`
}
