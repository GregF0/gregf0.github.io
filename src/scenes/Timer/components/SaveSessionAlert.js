import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import styled from 'styled-components'
import { useTheme } from '@material-ui/core'
import Dialog from '@material-ui/core/Dialog'
import DialogTitle from '@material-ui/core/DialogTitle'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogActions from '@material-ui/core/DialogActions'
import Button from '@material-ui/core/Button'
import { setSaveSessionAlert } from '../data/timer/actions'
import { startAddSession } from '../../../data/sessions/actions'

const Time = styled.span`
  color: ${({ color }) => color};
`

export const SaveSessionAlert = ({ time }) => {
  const { saveSessionAlert } = useSelector((state) => state.timer)
  const label = useSelector((state) => state.labels.labelSelected)

  const theme = useTheme()

  const dispatch = useDispatch()

  const closeAlert = () => {
    dispatch(setSaveSessionAlert(false))
  }

  const saveSession = () => {
    dispatch(
      startAddSession({
        duration: time,
        label: label ? label.id : null,
        createdAt: Date.now(),
      })
    )
    closeAlert()
  }

  return (
    <Dialog
      open={saveSessionAlert}
      onClose={closeAlert}
      aria-labelledby="save-session-alert-title"
      aria-describedby="save-session-alert-description"
    >
      <DialogTitle id="save-session-alert-title">Save session?</DialogTitle>
      <DialogContent>
        <DialogContentText id="save-session-alert-description">
          Do you want to save{' '}
          <Time color={theme.palette.secondary.main}>
            {time.minutes < 10 ? '0' + time.minutes : time.minutes}:
            {time.seconds < 10 ? '0' + time.seconds : time.seconds}
          </Time>{' '}
          to the statistics?
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={closeAlert}>Don't save</Button>
        <Button onClick={saveSession} color="primary">
          Save
        </Button>
      </DialogActions>
    </Dialog>
  )
}
