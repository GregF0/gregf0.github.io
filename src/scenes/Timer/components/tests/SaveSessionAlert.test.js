import React from 'react'
import * as redux from 'react-redux'
import { createShallow } from '@material-ui/core/test-utils'
import DialogContentText from '@material-ui/core/DialogContentText'
import Button from '@material-ui/core/Button'
import { SaveSessionAlert } from '../SaveSessionAlert'
import * as timerActions from '../../data/timer/actions'
import * as sessionsActions from '../../../../data/sessions/actions'

describe('<SaveSessionAlert />', () => {
  const time = { minutes: 4, seconds: 11 }
  const shallow = createShallow()
  const createWrapper = () => {
    return shallow(<SaveSessionAlert time={time} />)
  }

  const dispatchMocked = jest.fn()
  jest.spyOn(redux, 'useDispatch').mockImplementation(() => dispatchMocked)

  const createStore = () => {
    const store = {
      timer: { saveSessionAlert: true },
      labels: { labelSelected: null },
    }

    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((callback) => callback(store))

    return store
  }
  beforeEach(() => {
    jest.clearAllMocks()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  test('should render <SaveSessionAlert /> correctly', () => {
    createStore()
    expect(createWrapper()).toMatchSnapshot()
  })

  test('should render correct message', () => {
    expect(createWrapper().find(DialogContentText).text()).toBe(
      'Do you want to save 04:11 to the statistics?'
    )
  })

  test('should handle cancel button click', () => {
    const setSaveSessionAlertMocked = jest
      .spyOn(timerActions, 'setSaveSessionAlert')
      .mockImplementation(() => jest.fn())

    createStore()

    createWrapper().find(Button).at(0).simulate('click')
    expect(setSaveSessionAlertMocked).toHaveBeenCalledWith(false)
  })

  test('should call confirm button click', () => {
    const startAddSessionMocked = jest
      .spyOn(sessionsActions, 'startAddSession')
      .mockImplementation(() => jest.fn())

    createStore()

    createWrapper().find(Button).at(1).simulate('click')

    expect(startAddSessionMocked).toHaveBeenCalledWith({
      duration: time,
      label: null,
      createdAt: expect.any(Number),
    })
  })
})
