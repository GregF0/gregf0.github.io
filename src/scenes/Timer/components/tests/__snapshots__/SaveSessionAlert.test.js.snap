// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SaveSessionAlert /> should render <SaveSessionAlert /> correctly 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-describedby="save-session-alert-description"
  aria-labelledby="save-session-alert-title"
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="save-session-alert-title"
  >
    Save session?
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogContentText))
      id="save-session-alert-description"
    >
      Do you want to save
       
      <styled.span
        color="#f50057"
      >
        04
        :
        11
      </styled.span>
       
      to the statistics?
    </WithStyles(ForwardRef(DialogContentText))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      onClick={[Function]}
    >
      Don't save
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      onClick={[Function]}
    >
      Save
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
